consensus/istanbul/core/roundchange.go

```diff
@@ -32,79 +32,151 @@ func (c *core) sendNextRoundChange() {
-	cv = c.currentView()
-	rc := &istanbul.Subject{
-		View:   cv,
-		Digest: common.Hash{},
+	rc := &istanbul.RoundChange{
+		View:                nextView,
+		PreparedCertificate: c.current.preparedCertificate,
 	}
 
 	payload, err := Encode(rc)
 	if err != nil {
 		logger.Error("Failed to encode ROUND CHANGE", "rc", rc, "err", err)
 		return
 	}
-
-	c.broadcast(&message{
-		Code: msgRoundChange,
+	logger.Trace("Sending round change message", "rcs", c.roundChangeSet)
+	c.broadcast(&istanbul.Message{
+		Code: istanbul.MsgRoundChange,
 		Msg:  payload,
 	})
 }
 
-func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
-	logger := c.logger.New("state", c.state, "from", src.Address().Hex())
+func (c *core) handleRoundChangeCertificate(proposal istanbul.Subject, roundChangeCertificate istanbul.RoundChangeCertificate) error {
+	logger := c.logger.New("state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "handleRoundChangeCertificate")
+
+	if len(roundChangeCertificate.RoundChangeMessages) > c.valSet.Size() || len(roundChangeCertificate.RoundChangeMessages) < c.valSet.MinQuorumSize() {
+		return errInvalidRoundChangeCertificateNumMsgs
+	}
+
+	seen := make(map[common.Address]bool)
+	decodedMessages := make([]istanbul.RoundChange, len(roundChangeCertificate.RoundChangeMessages))
+	for i, message := range roundChangeCertificate.RoundChangeMessages {
+		// Verify message signed by a validator
+		data, err := message.PayloadNoSig()
+		if err != nil {
+			return err
+		}
+
+		signer, err := c.validateFn(data, message.Signature)
+		if err != nil {
+			return err
+		}
+
+		if signer != message.Address {
+			return errInvalidRoundChangeCertificateMsgSignature
+		}
+
+		// Check for duplicate ROUND CHANGE messages
+		if seen[signer] {
+			return errInvalidRoundChangeCertificateDuplicate
+		}
+		seen[signer] = true
+
+		// Check that the message is a ROUND CHANGE message
+		if istanbul.MsgRoundChange != message.Code {
+			return errInvalidRoundChangeCertificateMsgCode
+		}
+
+		var roundChange *istanbul.RoundChange
+		if err := message.Decode(&roundChange); err != nil {
+			logger.Error("Failed to decode ROUND CHANGE in certificate", "err", err)
+			return err
+		}
+
+		// Verify ROUND CHANGE message is for a proper view
+		// TODO(joshua): May be able to relax to the proposal is >= than the round change message.
+		if roundChange.View.Cmp(proposal.View) != 0 {
+			return errInvalidRoundChangeCertificateMsgView
+		}
+
+		if roundChange.HasPreparedCertificate() {
+			// Verify message for the proper proposal.
+			if proposal.Digest != roundChange.PreparedCertificate.Proposal.Hash() {
+				return errInvalidPreparedCertificateDigestMismatch
+			}
+
+			if err := c.verifyPreparedCertificate(roundChange.PreparedCertificate); err != nil {
+				return err
+			}
+		}
+
+		decodedMessages[i] = *roundChange
+		// TODO(joshua): startNewRound needs these round change messages to generate a
```

Comment:
https://github.com/celo-org/celo-blockchain/blob/trianglesphere/istanbul-liveness/consensus/istanbul/core/core.go#L256 checks if we can generate a prepared certificate (which requires 2f+1 round change messages). If it can't it assumes that it's waiting for a round change (even if it is trying to start a new round because it got a prepared certificate) which causes it to block.

Also part of the implicit round change state machine causing problems.

Reply:
My bad, `prepared certificate` should be `round change certificate`. Thanks for catching this.